### ISSUE TITLE
feat(macos): dock management api 

### ISSUE THREAD

**Issue by OceanBelongsToMe** [tauri-apps/tauri#2358]
### **Platform and Versions (required):**

```
Operating System

Node.js environment
Node.js 
@tauri-apps/cli 
@tauri-apps/api 

Global packages
npm 
yarn 

Rust environment
rustc 
cargo 

App directory structure


App
tauri.rs 
build-type 
CSP 
```

**Comment by probablykasper** [comment]
Hard to say what you mean because you didn't say much. Windows don't show up in the macOS Dock, so `window.set_skip_taskbar` isn't something that works on macOS. If you're talking about hiding the app from the Dock, that's mentioned in https://github.com/tauri-apps/tauri/issues/2258

**Comment by amrbashir** [comment]
It is not possible for macOS so I will rename this issue to dock management.

**Comment by amrbashir** [comment]
related to tauri-apps/tao#105

**Comment by betamos** [comment]
My use case, for whoever is looking at this in the future:

The user can control whether the app runs in (a) both foreground and background or (b) foreground only, using a checkbox:

- Tray icon is shown if and only if background mode is enabled.
- Dock icon is always shown when main window is open.
- Dock icon is hidden if main window is closed, under mode (a).

`set_activation_policy` almost solves my problem, but:

1. I can't seem to change it when the app is running, since it's exposed as `&mut self` on the `App` struct. (I'm not an expert on MacOS, but it appears this should be allowed, perhaps through the AppHandle?)
2. When set to `Accessory`, the dock icon is not shown, even if main window is open. (This appears to be MacOS default, so not expecting it to be overridden by Tauri).

**Comment by betamos** [comment]
Check out these commits here:

https://github.com/betamos/tauri/commits/betamos-head

(Relies on an unpublished addition to Tao)

I got everything to work except for one issue:

- User persists app in dock (either through "Show recent" or by adding it manually to the dock)
- Set activation policy to accessory/prohibited (disables icon, but doesn't remove it)
- Click icon (this doesn't *show* the app automatically, and there's no way to intercept the event atm, see https://github.com/tauri-apps/tao/issues/218)

It works well otherwise, and everything I added is platform-gated with `cfg(target_os="macos")`. It's basically a thin layer on top of MacOS APIs. I'd be happy to contribute but someone should probably vet/approve/feedback the approach first.

## Alternative approach

Instead of using archaic terminology such as `set_activation_policy`, I'd prefer adding simpler APIs that exist on all platforms to avoid having to fiddle with platform-gates in business logic. (I suspect a verbose above approach – such as the one above – will result in cargo-cult copy-pasting, so why not make it simpler from the get-go?). Example:

- show/hide_dock_icon
- show/hide_application
- or possibly combined in a show_all/hide_all with one or two params that does the right thing for all platforms - in most cases

There's some precedent for this in `skip_taskbar`, which doesn't apply to MacOS.

**Comment by probablykasper** [comment]
`set_activation_policy` is an entirely different thing from showing/hiding the dock icon - Keep in mind there are three different activation policies. If you change the activation policy, you should be aware of what it does, so renaming it could potentially be misleading

**Comment by betamos** [comment]
Yes, I think only `regular` and `accessory` would apply here. Is there any API you would suggest for dock icon control? I only found `transformProcessType` (which is used by Electron) but it seems racey judging by their workarounds.

https://github.com/electron/electron/blob/b0f315a637b05bc445fc7e6084baa840891c1207/shell/browser/browser_mac.mm#L414

**Comment by probablykasper** [comment]
People have been mentioning `transformProcessType`, and it seems like the only option as far as I can tell

**Comment by PavelLaptev** [comment]
It would be nice to be able to dynamically change the dock icon 🙂

**Comment by thewh1teagle** [comment]
Hey
What's the status of this feature?
